fe2-android/app/src/test/java/com/github/dedis/student20_pop/model/data/LAORepositoryTest.java

```diff
@@ -0,0 +1,156 @@
+package com.github.dedis.student20_pop.model.data;
+
+import com.github.dedis.student20_pop.Injection;
+import com.github.dedis.student20_pop.model.Lao;
+import com.github.dedis.student20_pop.model.network.GenericMessage;
+import com.github.dedis.student20_pop.model.network.answer.Answer;
+import com.github.dedis.student20_pop.model.network.answer.Result;
+import com.github.dedis.student20_pop.model.network.method.message.MessageGeneral;
+import com.github.dedis.student20_pop.utility.scheduler.SchedulerProvider;
+import com.github.dedis.student20_pop.utility.scheduler.TestSchedulerProvider;
+import com.google.crypto.tink.integration.android.AndroidKeysetManager;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+import io.reactivex.observers.TestObserver;
+import io.reactivex.schedulers.TestScheduler;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import junit.framework.TestCase;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class LAORepositoryTest extends TestCase {
+
+  @Mock
+  LAORemoteDataSource remoteDataSource;
+
+  @Mock
+  LAOLocalDataSource localDataSource;
+
+  @Mock
+  AndroidKeysetManager androidKeysetManager;
+
+  @Mock
+  MessageGeneral messageGeneral;
+
+  private TestScheduler testScheduler;
+  private LAORepository repository;
+
+  @Before
+  public void setup() {
+    SchedulerProvider testSchedulerProvider = new TestSchedulerProvider();
+    testScheduler = (TestScheduler) testSchedulerProvider.io();
+
+    // Simulate a network response from the server after 1 second
+    Observable<GenericMessage> upstream = Observable.just((GenericMessage) new Result(42))
+        .delay(1, TimeUnit.SECONDS, testScheduler);
+
+    Mockito.when(remoteDataSource.observeMessage()).thenReturn(upstream);
+    Mockito.when(remoteDataSource.observeWebsocket()).thenReturn(Observable.empty());
+    Mockito.when(remoteDataSource.incrementAndGetRequestId()).thenReturn(42);
+    repository = LAORepository
+        .getInstance(remoteDataSource, localDataSource, androidKeysetManager,
+            Injection.provideGson(), testSchedulerProvider);
+  }
+
+  @After
+  public void destroy() {
+    LAORepository.destroyInstance();
+  }
+
+  @Test
+  public void testSendCatchup() {
+    Single<Answer> answerCatchup = repository.sendCatchup("/root/");
+    Mockito.verify(remoteDataSource, Mockito.times(1)).sendMessage(Mockito.any());
+
+    TestObserver<Answer> testObserverCatchup = TestObserver.create();
+
+    answerCatchup.subscribe(testObserverCatchup);
+
+    testScheduler.advanceTimeBy(950, TimeUnit.MILLISECONDS);
+    testObserverCatchup.assertNotComplete();
+
+    testScheduler.advanceTimeBy(100, TimeUnit.MILLISECONDS);
```

Comment:
assuming we have a `RESPONSE_DELAY` constant set at 1000 ms

```suggestion
    testScheduler.advanceTimeTo(RESPONSE_DELAY-1, TimeUnit.MILLISECONDS);
    testObserverCatchup.assertNotComplete();

    testScheduler.advanceTimeTo(RESPONSE_DELAY+1, TimeUnit.MILLISECONDS);
```